moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+      p = doPeek()
+    }
+    if (p == PEEKED_BEGIN_ARRAY) {
+      pushScope(JsonScope.EMPTY_ARRAY)
+      pathIndices[stackSize - 1] = 0
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected BEGIN_ARRAY but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun endArray() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    peeked = if (p == PEEKED_END_ARRAY) {
+      stackSize--
+      pathIndices[stackSize - 1]++
+      PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected END_ARRAY but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun beginObject() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    peeked = if (p == PEEKED_BEGIN_OBJECT) {
+      pushScope(JsonScope.EMPTY_OBJECT)
+      PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected BEGIN_OBJECT but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun endObject() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_END_OBJECT) {
+      stackSize--
+      pathNames[stackSize] = null // Free the last path name so that it can be garbage collected!
+      pathIndices[stackSize - 1]++
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected END_OBJECT but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun hasNext(): Boolean {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    return p != PEEKED_END_OBJECT && p != PEEKED_END_ARRAY && p != PEEKED_EOF
+  }
+
+  @Throws(IOException::class)
+  override fun peek(): Token {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    return when (p) {
+      PEEKED_BEGIN_OBJECT -> Token.BEGIN_OBJECT
+      PEEKED_END_OBJECT -> Token.END_OBJECT
+      PEEKED_BEGIN_ARRAY -> Token.BEGIN_ARRAY
+      PEEKED_END_ARRAY -> Token.END_ARRAY
+      PEEKED_SINGLE_QUOTED_NAME, PEEKED_DOUBLE_QUOTED_NAME, PEEKED_UNQUOTED_NAME, PEEKED_BUFFERED_NAME -> Token.NAME
+      PEEKED_TRUE, PEEKED_FALSE -> Token.BOOLEAN
+      PEEKED_NULL -> Token.NULL
+      PEEKED_SINGLE_QUOTED, PEEKED_DOUBLE_QUOTED, PEEKED_UNQUOTED, PEEKED_BUFFERED -> Token.STRING
+      PEEKED_LONG, PEEKED_NUMBER -> Token.NUMBER
+      PEEKED_EOF -> Token.END_DOCUMENT
+      else -> throw AssertionError()
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun doPeek(): Int {
+    val peekStack = scopes[stackSize - 1]
+    when (peekStack) {
+      JsonScope.EMPTY_ARRAY -> scopes[stackSize - 1] = JsonScope.NONEMPTY_ARRAY
+      JsonScope.NONEMPTY_ARRAY -> {
+        // Look for a comma before the next element.
+        val c = nextNonWhitespace(true).toChar()
+        buffer.readByte() // consume ']' or ','.
+        when (c) {
+          ']' -> return PEEKED_END_ARRAY.also { peeked = it }
```

Comment:
the original Java sources for this were: 

```java
return peeked = PEEKED_END_ARRAY
```

I figured this is the equivalent Kotlin, but I'm also happy to convert it to 

```kotlin
peeked = PEEKED_END_ARRAY
return peeked
```
if the `.also` construction doesn't feel readable/idiomatic/maintainable

Reply:
I prefer the split version